lib/spack/spack/test/cmd/location.py

```diff
@@ -0,0 +1,116 @@
+# Copyright 2013-2019 Lawrence Livermore National Security, LLC and other
+# Spack Project Developers. See the top-level COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+import os
+import shutil
+import pytest
+
+from llnl.util.filesystem import mkdirp
+import spack.environment as ev
+from spack.cmd.location import BadEnvironmentError, MissingBuildDirectoryError
+from spack.cmd.location import SpecificationError
+from spack.main import SpackCommand
+import spack.paths
+
+
+# everything here uses the mock_env_path
+pytestmark = pytest.mark.usefixtures('config')
+
+# location prints out "locations of packages and spack directories"
+location = SpackCommand('location')
+
+_repo_dir = os.path.join(spack.paths.repos_path, 'builtin.mock')
+_pkg_dir = os.path.join(_repo_dir, 'packages')
+
+_mock_spec_name = 'externaltest'
+_test_env_name = 'test'
+
+
+@pytest.fixture
+def mock_test_env(mutable_mock_env_path):
+    env_dir = ev.root(_test_env_name)
+    mkdirp(env_dir)
+    yield
+
+    shutil.rmtree(env_dir)
+
+
+@pytest.fixture
+def mock_spec():
+    s = spack.spec.Spec(_mock_spec_name).concretized()
+    pkg = spack.repo.get(s)
+
+    # Make it look like the source was actually expanded.
+    source_path = pkg.stage.source_path
+    mkdirp(source_path)
+    yield
+
+    # The test is done so be sure to remove the spec from the mock stage area.
+    shutil.rmtree(pkg.stage.path)
+
+
+def test_location_bad_env():
+    """Tests spack location with an unknown environment."""
+    with pytest.raises(BadEnvironmentError, match="No such environment"):
+        location('--env', 'test-unknown-env')
+
+
+@pytest.mark.parametrize('spec1,spec2,msg', [
+    (None, None, "You must supply a spec"),
+    ('spec1', 'spec2', "Only one spec, not 2")])
+def test_location_bad_spec_options(spec1, spec2, msg):
+    """Tests spack location with bad spec options."""
+    args = [spec1, spec2]
+    if not all(args):
+        args = []
+
+    with pytest.raises(SpecificationError, match=msg):
+        location(*args)
+
+
+@pytest.mark.db
+@pytest.mark.usefixtures('database', 'mock_spec')
+def test_location_build_dir():
+    """Tests spack location --build-dir"""
+    out = location('--build-dir', _mock_spec_name)
+    assert _mock_spec_name in out
```

Comment:
IMO it would be good to work with the spec here (i.e. do `yield s` in `mock_spec`), and then retrieve the directory you expect the `location` command to output from the `Spec` object. I think that this would have the following benefits:

* More-complete comparison (since you can compare the full path without reconstructing it)
* Avoids replicating knowledge of the structure of the `Stage` object etc. outside of the `Stage`

Likewise for the testing of output in `test_location_env`. I feel similarly for the `--install-dir`/`--stage-dir` checks in `test_location_options`  but I realize you get a significant amount of coverage with a limited amount of work there.

Reply:
I agree in principle with everything Peter said, but just remember to balance the effort put in with the coverage we're getting out.  I wouldn't go to great lengths to implement new test scaffolding here.  Keep the tests for location simple.

Reply:
> I agree in principle with everything Peter said, but just remember to balance the effort put in with the coverage we're getting out. I wouldn't go to great lengths to implement new test scaffolding here. Keep the tests for location simple.

I understand though, since I'm new to PyTest, it is a good opportunity for me to dig in and learn more about it.